src/apply_rules/object_rules.rs

```diff
@@ -0,0 +1,508 @@
+// Copyright (C) 2017 Kisio Digital and/or its affiliates.
+//
+// This program is free software: you can redistribute it and/or modify it
+// under the terms of the GNU Affero General Public License as published by the
+// Free Software Foundation, version 3.
+
+// This program is distributed in the hope that it will be useful, but WITHOUT
+// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+// FOR A PARTICULAR PURPOSE. See the GNU Affero General Public License for more
+// details.
+
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>
+
+use crate::{
+    model::{Collections, CO2_EMISSIONS},
+    objects::{
+        CommercialMode, Line, Network, ObjectType as ModelObjectType, PhysicalMode, VehicleJourney,
+    },
+    utils::{Report, ReportType},
+    Result,
+};
+use failure::format_err;
+use log::info;
+use relational_types::IdxSet;
+use serde::Deserialize;
+use serde_json::{error::Category, error::Error as SerdeJsonError, Value};
+use std::{
+    collections::{HashMap, HashSet},
+    fs::File,
+    path::Path,
+    result::Result as StdResult,
+};
+
+#[derive(Debug, Deserialize)]
+struct ObjectProperties {
+    #[serde(flatten)]
+    properties: HashMap<String, Value>,
+    #[serde(default)]
+    grouped_from: Vec<String>,
+}
+
+fn get_value_string_from_properties(
+    property: &serde_json::value::Value,
+    key: &str,
+    default_value: &str,
+) -> String {
+    if let Some(value) = property.get(key) {
+        String::deserialize(value)
+            .map(|opt| opt)
+            .unwrap_or_else(|_| default_value.to_string())
+    } else {
+        default_value.to_string()
+    }
+}
+
+fn get_opt_value_string_from_properties(
+    property: &serde_json::value::Value,
+    key: &str,
+    default_value: Option<String>,
+) -> Option<String> {
+    if let Some(value) = property.get(key) {
+        Option::<String>::deserialize(value)
+            .map(|opt| opt)
+            .unwrap_or(default_value)
+    } else {
+        default_value
+    }
+}
+
+fn get_opt_value_number_from_properties(
+    property: &serde_json::value::Value,
+    key: &str,
+    default_value: Option<u32>,
+) -> Option<u32> {
+    if let Some(value) = property.get(key) {
+        Option::<u32>::deserialize(value)
+            .map(|opt| opt)
+            .unwrap_or(default_value)
+    } else {
+        default_value
+    }
+}
```

Comment:
These 2 functions are identical. I believe they can be factorized with a generic.

```rust
fn get_opt_value_from_properties<T>(
    property: &serde_json::value::Value,
    key: &str,
    default_value: Option<T>,
) -> Option<T>;
```